### PR TITLE
Process task/workflows in skupack

### PR DESCRIPTION
**Pull request by zyoung51**
- Add workflow/task bundling into skupack.  The 'taskRoot' and 'workflowRoot'
  properties can be added to the config.json to specify directories containing
  tasks and graphs to be loaded as part of the skupack.  The tasks and graphs
  will have '::skuid' appended to scope them to the sku.  If tasks are
  included in the skupack and referenced in the graph, then the task names in
  the graph will be updated to the ::skuid scoped names.

### CODE REVIEW

**Comment by zyoung51** [comment]
@RackHD/corecommitters @uppalk1 @keedya @tannoa2

**Comment by zyoung51** [comment]
test this please

File: lib/services/sku-pack-service.js [hunk 1]
```diff
@@ -43,6 +45,24 @@ function skuPackServiceFactory(
     }
 
     /**
+     * Given a nodeId, return the prioritized scope order
+     * @param  {string}     nodeId
+     */    
+     SkuPackService.prototype.setupScope = function(nodeId) {
+        var defaultScope = ['global'];
+        if(typeof(nodeId) === 'undefined')  {
```
Comment (benbp) [inline lib/services/sku-pack-service.js]: Why not just `if(!nodeId)`? This way if nodeId is null the conditional still hits.

File: lib/services/sku-pack-service.js [hunk 2]
```diff
@@ -97,11 +115,43 @@ function skuPackServiceFactory(
...
+                var tasks = [];
+                var taskPromises = Promise.resolve([]);
+                if(conf.hasOwnProperty('taskRoot')) {
+                    var taskRoot = skuRoot + path.resolve('/', conf.taskRoot);
+                    taskPromises = loadWorkflowItems(skuName, taskRoot, waterline.taskdefinitions);
+                    taskPromises = taskPromises.map(function(contents) {
+                        var data = JSON.parse(contents);
+                        tasks.push(data.injectableName);
+                        data.injectableName = data.injectableName + '::' + skuName;
+                        taskgraphRunner.defineTask(data);
```
Comment (benbp) [inline lib/services/sku-pack-service.js]: This needs to be returned?

Also, more rebase hell for me :P

File: lib/services/sku-pack-service.js [hunk 3]
```diff
@@ -97,11 +115,43 @@ function skuPackServiceFactory(
...
+                        return loadWorkflowItems(skuName, workflowRoot, waterline.graphdefinitions);
+                    }).map(function(contents) {
+                        var data = JSON.parse(contents);
+                        data.injectableName = data.injectableName + '::' + skuName;
+                        _.forEach(getObjectsWithKey(data,'taskName'), function(item) {
+                            if(tasks.indexOf(item.taskName) !== -1) {
+                                item.taskName = item.taskName + '::' + skuName;
+                            }
+                        });
+                        taskgraphRunner.defineTaskGraph(data);
```
Comment (benbp) [inline lib/services/sku-pack-service.js]: Same, needs to be returned.

File: lib/services/sku-pack-service.js [hunk 4]
```diff
@@ -271,5 +337,50 @@ function skuPackServiceFactory(
...
+                return fs.statSync(fromRoot + '/' + entry).isFile();
+            })
+            .map(function(entry) {
+                return fs.readFileAsync(fromRoot + '/' + entry);
+            })
+            .filter(function(contents) {
+                var data = JSON.parse(contents);
+                return dbCatalog.findOne({injectableName: data.injectableName + '::' + skuName})
+                    .then(function(item) {
+                        return (typeof(item) === 'undefined') ? true : false;
```
Comment (benbp) [inline lib/services/sku-pack-service.js]: Isn't waterline going to return `null` here? Which will pass this check because `typeof(null) === 'object'`

**Comment by jlongever** [comment]
test this please

File: lib/services/sku-pack-service.js (hunk 4 above)
Comment (zyoung51) [inline lib/services/sku-pack-service.js]: I was seeing undefined there which led to that check.  I'll double check.

**Comment by jlongever** [comment]
test this please

**Comment by jlongever** [comment]
+1

**Comment by benbp** [comment]
:+1:

**Comment by benbp** [comment]
Testing

- [ ] do this thing

**Comment by jlongever** [comment]
+1 again

**Pull request merged**